File: docx-core/src/documents/elements/table.rs
use serde::ser::{SerializeStruct, Serializer};
use serde::Serialize;

use super::*;
use crate::documents::BuildXML;
use crate::types::*;
use crate::xml_builder::*;

#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Table {
    pub rows: Vec<TableChild>,
    pub grid: Vec<usize>,
    pub has_numbering: bool,
    pub property: TableProperty,
}

#[derive(Debug, Clone, PartialEq)]
pub enum TableChild {
    TableRow(TableRow),
}

impl BuildXML for TableChild {
    fn build(&self) -> Vec<u8> {
        match self {
            TableChild::TableRow(v) => v.build(),
        }
    }
}

impl Table {
    pub fn new(rows: Vec<TableRow>) -> Table {
        let property = TableProperty::new();
        let has_numbering = rows.iter().any(|c| c.has_numbering);
        let grid = vec![];
        let rows = rows.into_iter().map(TableChild::TableRow).collect();
        Self {
            property,
            rows,
            grid,
            has_numbering,
        }
    }

    pub fn without_borders(rows: Vec<TableRow>) -> Table {
        let property = TableProperty::without_borders();
        let has_numbering = rows.iter().any(|c| c.has_numbering);
        let grid = vec![];
        let rows = rows.into_iter().map(TableChild::TableRow).collect();
        Self {
            property,
            rows,
            grid,
            has_numbering,
        }
    }

    pub fn add_row(mut self, row: TableRow) -> Table {
        self.rows.push(TableChild::TableRow(row));
        self
    }

    pub fn set_grid(mut self, grid: Vec<usize>) -> Table {
        self.grid = grid;
        self
    }

    pub fn indent(mut self, v: i32) -> Table {
        self.property = self.property.indent(v);
        self
    }

    pub fn align(mut self, v: TableAlignmentType) -> Table {
        self.property = self.property.align(v);
        self
    }

    pub fn style(mut self, s: impl Into<String>) -> Table {
        self.property = self.property.style(s);
        self
    }

    pub fn layout(mut self, t: TableLayoutType) -> Table {
        self.property = self.property.layout(t);
        self
    }

    pub fn position(mut self, p: TablePositionProperty) -> Self {
        self.property = self.property.position(p);
        self
    }

    pub fn width(mut self, w: usize, t: WidthType) -> Table {
        self.property = self.property.width(w, t);
        self
    }

    pub fn margins(mut self, margins: TableCellMargins) -> Self {
        self.property = self.property.set_margins(margins);
        self
    }

    pub fn set_borders(mut self, borders: TableBorders) -> Self {
        self.property = self.property.set_borders(borders);
        self
    }

    pub fn set_border(mut self, border: TableBorder) -> Self {
        self.property = self.property.set_border(border);
        self
    }

    pub fn clear_border(mut self, position: TableBorderPosition) -> Self {
        self.property = self.property.clear_border(position);
        self
    }

    pub fn clear_all_border(mut self) -> Self {
        self.property = self.property.clear_all_border();
        self
    }
}

impl BuildXML for Table {
    fn build(&self) -> Vec<u8> {
        let grid = TableGrid::new(self.grid.clone());
        let b = XMLBuilder::new()
            .open_table()
            .add_child(&self.property)
            .add_child(&grid)
            .add_children(&self.rows);
        b.close().build()
    }
}

impl BuildXML for Box<Table> {
    fn build(&self) -> Vec<u8> {
        Table::build(self)
    }
}

impl Serialize for TableChild {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        match *self {
            TableChild::TableRow(ref r) => {
                let mut t = serializer.serialize_struct("TableRow", 2)?;
                t.serialize_field("type", "tableRow")?;
                t.serialize_field("data", r)?;
                t.end()
            }
        }
    }
}

#[cfg(test)]
mod tests {

    use super::*;
    #[cfg(test)]
    use pretty_assertions::assert_eq;
    use std::str;

    #[test]
    fn test_table() {
        let b = Table::new(vec![TableRow::new(vec![])]).build();
        assert_eq!(
            str::from_utf8(&b).unwrap(),
            r#"<w:tbl><w:tblPr><w:tblW w:w="0" w:type="auto" /><w:jc w:val="left" /><w:tblBorders><w:top w:val="single" w:sz="2" w:space="0" w:color="000000" /><w:left w:val="single" w:sz="2" w:space="0" w:color="000000" /><w:bottom w:val="single" w:sz="2" w:space="0" w:color="000000" /><w:right w:val="single" w:sz="2" w:space="0" w:color="000000" /><w:insideH w:val="single" w:sz="2" w:space="0" w:color="000000" /><w:insideV w:val="single" w:sz="2" w:space="0" w:color="000000" /></w:tblBorders></w:tblPr><w:tblGrid /><w:tr><w:trPr /></w:tr></w:tbl>"#
        );
    }

    #[test]
    fn test_table_grid() {
        let b = Table::new(vec![TableRow::new(vec![])])
            .set_grid(vec![100, 200])
            .build();
        assert_eq!(
            str::from_utf8(&b).unwrap(),
            r#"<w:tbl><w:tblPr><w:tblW w:w="0" w:type="auto" /><w:jc w:val="left" /><w:tblBorders><w:top w:val="single" w:sz="2" w:space="0" w:color="000000" /><w:left w:val="single" w:sz="2" w:space="0" w:color="000000" /><w:bottom w:val="single" w:sz="2" w:space="0" w:color="000000" /><w:right w:val="single" w:sz="2" w:space="0" w:color="000000" /><w:insideH w:val="single" w:sz="2" w:space="0" w:color="000000" /><w:insideV w:val="single" w:sz="2" w:space="0" w:color="000000" /></w:tblBorders></w:tblPr><w:tblGrid>
  <w:gridCol w:w="100" w:type="dxa" />
  <w:gridCol w:w="200" w:type="dxa" />
</w:tblGrid><w:tr><w:trPr /></w:tr></w:tbl>"#
        );
    }

    #[test]
    fn test_table_json() {
        let t = Table::new(vec![]).set_grid(vec![100, 200, 300]);
        assert_eq!(
            serde_json::to_string(&t).unwrap(),
            r#"{"rows":[],"grid":[100,200,300],"hasNumbering":false,"property":{"width":{"width":0,"widthType":"auto"},"justification":"left","borders":{"top":{"borderType":"single","size":2,"color":"000000","position":"top","space":0},"left":{"borderType":"single","size":2,"color":"000000","position":"left","space":0},"bottom":{"borderType":"single","size":2,"color":"000000","position":"bottom","space":0},"right":{"borderType":"single","size":2,"color":"000000","position":"right","space":0},"insideH":{"borderType":"single","size":2,"color":"000000","position":"insideH","space":0},"insideV":{"borderType":"single","size":2,"color":"000000","position":"insideV","space":0}}}}"#
        );
    }
}

File: docx-wasm/js/table.ts
import * as wasm from "./pkg";

import { WidthType } from ".";
import { TableRow } from "./table-row";

export type TableAlignmentType = "center" | "left" | "right";
export type TableLayoutType = "fixed" | "autofit";

export type TableProperty = {
  indent?: number;
  align?: TableAlignmentType;
  width?: number;
  styleId?: string;
  cellMargins: {
    top: { val: number; type: WidthType };
    left: { val: number; type: WidthType };
    bottom: { val: number; type: WidthType };
    right: { val: number; type: WidthType };
  };
  layout?: TableLayoutType;
};

export const createDefaultTableCellMargins = () => {
  return {
    top: { val: 0, type: "dxa" },
    left: { val: 55, type: "dxa" },
    bottom: { val: 0, type: "dxa" },
    right: { val: 55, type: "dxa" },
  } as const;
};

export class Table {
  hasNumberings = false;
  rows: TableRow[] = [];
  grid: number[] = [];
  property: TableProperty = {
    cellMargins: createDefaultTableCellMargins(),
  };

  addRow(row: TableRow) {
    if (row.hasNumberings) {
      this.hasNumberings = true;
    }
    this.rows.push(row);
    return this;
  }

  style(id: string) {
    this.property.styleId = id;
    return this;
  }

  setGrid(grid: number[]) {
    this.grid = grid;
    return this;
  }

  indent(v: number) {
    this.property.indent = v;
    return this;
  }

  align(v: TableAlignmentType) {
    this.property.align = v;
    return this;
  }

  layout(l: TableLayoutType) {
    this.property.layout = l;
    return this;
  }

  width(w: number) {
    this.property.width = w;
    return this;
  }

  cellMargins(top: number, right: number, bottom: number, left: number) {
    this.property.cellMargins = {
      top: { val: top, type: "dxa" },
      left: { val: left, type: "dxa" },
      bottom: { val: bottom, type: "dxa" },
      right: { val: right, type: "dxa" },
    };
    return this;
  }

  cellMarginTop(v: number, t: WidthType) {
    this.property.cellMargins.top = { val: v, type: t };
    return this;
  }

  cellMarginLeft(v: number, t: WidthType) {
    this.property.cellMargins.left = { val: v, type: t };
    return this;
  }

  cellMarginRight(v: number, t: WidthType) {
    this.property.cellMargins.right = { val: v, type: t };
    return this;
  }

  cellMarginBottom(v: number, t: WidthType) {
    this.property.cellMargins.bottom = { val: v, type: t };
    return this;
  }
}

export const convertWidthType = (t: string) => {
  switch (t) {
    case "nil":
    case "Nil":
      return wasm.WidthType.Nil;
    case "Pct":
    case "pct":
      return wasm.WidthType.Pct;
    case "DXA":
    case "dxa":
      return wasm.WidthType.Dxa;
    case "Auto":
    case "auto":
      return wasm.WidthType.Auto;
    default:
      return wasm.WidthType.Dxa;
  }
};

export const setTableProperty = <T extends wasm.Table | wasm.Style>(
  target: T,
  property: TableProperty
): T => {
  if (target instanceof wasm.Table) {
    target = target.indent(property.indent ?? 0) as T;
  } else if (target instanceof wasm.Style) {
    target = target.table_indent(property.indent ?? 0) as T;
  }

  if (property.cellMargins) {
    const { top, right, bottom, left } = property.cellMargins;
    target = target
      .cell_margin_top(top.val, convertWidthType(top.type))
      .cell_margin_right(right.val, convertWidthType(right.type))
      .cell_margin_bottom(bottom.val, convertWidthType(bottom.type))
      .cell_margin_left(left.val, convertWidthType(left.type)) as T;
  }

  const align = ((): wasm.TableAlignmentType | null => {
    switch (property.align) {
      case "center": {
        return wasm.TableAlignmentType.Center;
      }
      case "right": {
        return wasm.TableAlignmentType.Right;
      }
      case "left": {
        return wasm.TableAlignmentType.Left;
      }
      default:
        return null;
    }
  })();

  if (align != null) {
    if (target instanceof wasm.Table) {
      target = target.align(align) as T;
    } else if (target instanceof wasm.Style) {
      target = target.table_align(align) as T;
    }
  }

  switch (property.layout) {
    case "fixed": {
      target = target.layout(wasm.TableLayoutType.Fixed) as T;
      break;
    }
    case "autofit": {
      target = target.layout(wasm.TableLayoutType.Autofit) as T;
      break;
    }
  }
  return target;
};
